servlet-engine/src/main/java/io/micronaut/servlet/engine/DefaultServletHttpRequest.java

```diff
@@ -147,74 +141,11 @@ protected DefaultServletHttpRequest(ConversionService conversionService,
         this.parameters = new ServletParameters();
         this.response = new DefaultServletHttpResponse<>(conversionService, this, response);
         this.body = SupplierUtil.memoizedNonEmpty(() -> {
-            B built = (B) buildBody();
+            B built = (B) bodyBuilder.buildBody(this::getInputStream, this);
             return Optional.ofNullable(built);
         });
     }
 
-    @Nullable
-    protected Object buildBody() {
-        final MediaType contentType = getContentType().orElse(MediaType.APPLICATION_JSON_TYPE);
-        if (isFormSubmission(contentType)) {
-            return getParameters().asMap();
-        } else {
-            if (delegate.getContentLength() == 0) {
```

Comment:
the old code checked against 0 which I think it is incorrect. I have changed to check against -1

Reply:
i think 0 might be correct. in netty there's a bunch of places where for content length 0, we have a special no-value case.